GCM encryption stream

ggershinsky opened the pull request:
Implements #2060

Besides encrypting and signing metadata (avro, json, puffin), this stream can also encrypt Avro data.

liujinhui1994 commented:
@ggershinsky  Is this PR still going on?

ggershinsky commented:
same as in https://github.com/apache/iceberg/pull/2639#issuecomment-1006487455

ggershinsky commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,201 @@
+import static org.apache.iceberg.encryption.AesGcmOutputStream.HEADER_SIZE_LENGTH;
+
+public class AesGcmInputStream extends SeekableInputStream {
+  private SeekableInputStream sourceStream;
```
sure

ggershinsky commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,201 @@
+  private long plainStreamSize;
+  private byte[] fileAadPrefix;
+
+  AesGcmInputStream(SeekableInputStream sourceStream, long sourceLength,
```
Always created by an `AesGcmInputFile` in the same package.

ggershinsky commented:
> Thanks @ggershinsky for the patch. Can we add a unit test for these two classes?

done

flyrain commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputFile.java:
```diff
@@ -0,0 +1,67 @@
+
+  @Override
+  public long getLength() {
+    Preconditions.checkArgument(plaintextLength >= 0, "Length is known after new stream is created");
```
Should we set the length if field plaintextLength is -1? Looks like this method is valid only after `newStream()` is called.
```
if(plaintextLength == -1) {
this.newStream();
}
```

RussellSpitzer replied:
When do we close the stream in this case? Or do we just make and close it in that case?

ggershinsky replied:
Both suggestions sound good to me.

flyrain commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,212 @@
+  @Override
+  public void seek(long newPos) throws IOException {
+    if (newPos >= plainStreamSize) {
+      throw new IOException("At or beyond max stream size " + plainStreamSize + ", " + newPos);
```
The massage could be
```
"The seeking position " + newPos + " has reached or exceeded the max stream size "  + plainStreamSize.
```

flyrain commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,212 @@
+
+  @Override
+  public synchronized void reset() throws IOException {
+    throw new UnsupportedOperationException();
```
Looks like we don't have to implement it, the abstract class `InputStream` has the same logic.

flyrain commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmOutputFile.java:
```diff
@@ -0,0 +1,64 @@
+    try {
+      return new AesGcmOutputStream(targetFile.create(), dataKey, null);
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
```
Can we add message like, "Failed to create file: %s", targetFile.location()?

flyrain commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmOutputFile.java:
```diff
@@ -0,0 +1,64 @@
+    try {
+      return new AesGcmOutputStream(targetFile.createOrOverwrite(), dataKey, null);
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
```
Can we add message like, "Failed to create or overwrite file: %s", targetFile.location()?

flyrain commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmOutputStream.java:
```diff
@@ -0,0 +1,173 @@
+  private SecretKey key;
+  private byte[] nonce;
+
+  private int blockSize = 1024 * 1024;
```
Make it a final static field?

flyrain commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmOutputStream.java:
```diff
@@ -0,0 +1,173 @@
+  private int positionInBuffer;
+  private long streamPosition;
+  private int currentBlockIndex;
+  private byte[] fileAadPrefix;
```
It can be a final field.

flyrain commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmOutputStream.java:
```diff
@@ -0,0 +1,173 @@
+    try {
+      int encrypted = gcmCipher.doFinal(plaintextBlockBuffer, 0, positionInBuffer, cipherText, GCM_NONCE_LENGTH);
+      Preconditions.checkArgument((encrypted == (positionInBuffer + GCM_TAG_LENGTH)),
+          "Wrong length of encrypted output: " + encrypted + " vs " + (positionInBuffer + GCM_TAG_LENGTH));
```
How about a message like this?
```
"Wrong length of encrypted output:  expected " + (positionInBuffer + GCM_TAG_LENGTH)) + " but " + encrypted;
```

flyrain commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmOutputStream.java:
```diff
@@ -0,0 +1,173 @@
+    targetStream.write(cipherText);
+  }
+
+  static byte[] calculateAAD(byte[] fileAadPrefix, int currentBlockIndex) {
```
This can go to a common place as well since both classes need it. We may have a util class, e.g. `GcmUtil`

flyrain commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,212 @@
+    try {
+      gcmCipher = Cipher.getInstance("AES/GCM/NoPadding");
+    } catch (GeneralSecurityException e) {
+      throw new IOException(e);
```
Add message for easier debugging?
```
new IOException("Failed to get an instance of GCM cipher",  e)
```

flyrain commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,212 @@
+    this.sourceStream = sourceStream;
+    byte[] prefixBytes = new byte[AesGcmOutputStream.PREFIX_LENGTH];
+    int fetched = sourceStream.read(prefixBytes);
+    Preconditions.checkArgument(fetched == AesGcmOutputStream.PREFIX_LENGTH,
+        "Insufficient read " + fetched);
```
They can be in the same line.
IIUC, this means the stream length is not long enough. Can we say something like "The stream length should be at least " + AesGcmOutputStream.PREFIX_LENGTH?

RussellSpitzer replied:
Aren't we also in danger of the read being less than prefix bytes?

flyrain commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,201 @@
+  private int numberOfBlocks;
+  private int lastBlockSize;
+  private long plainStreamSize;
+  private byte[] fileAadPrefix;
```
Both plainStreamSize and fileAadPrefix could be final.

flyrain commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,212 @@
+  private long netSourceFileSize;
+
+  private Cipher gcmCipher;
+  private SecretKey key;
+  private byte[] nonce;
```
final as well.

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,212 @@
+    System.arraycopy(prefixBytes, 0, magic, 0, AesGcmOutputStream.MAGIC_ARRAY.length);
+
+    Preconditions.checkArgument(Arrays.equals(AesGcmOutputStream.MAGIC_ARRAY, magic),
+        "File with wrong magic string. Should start with " + AesGcmOutputStream.MAGIC_STRING);
```
Instead of "file with wrong magic string" -> "Cannot read unencrypted file, missing header containing: ..."?

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,212 @@
+      throw new IOException(e);
+    }
+    this.nonce = new byte[AesGcmOutputStream.GCM_NONCE_LENGTH];
+    this.key = new SecretKeySpec(aesKey, "AES");
```
Constant?

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,212 @@
+    this.currentBlockIndex = 0;
+    this.currentOffsetInPlainBlock = 0;
+
+    numberOfBlocks = (int) (netSourceFileSize / cipherBlockSize);
```
Can we make the math here a little more explicit, all of the -1 and such make things a little hard for me to read. We discussed this a bit before offline, I mainly just don't want to have to think about why we have a +1 here or why we have a -1 somewhere else etc ...

I think we were discussing a something like "number of full blocks", "remainder" etc ...

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,212 @@
+    sourceStream.seek(AesGcmOutputStream.PREFIX_LENGTH + currentBlockIndex * cipherBlockSize);
+
+    while (remaining > 0) {
+      int toLoad = lastBlock ? lastBlockSize : cipherBlockSize;
```
This is part of the code I think is kind of complicated because of the special casing of the last block logic. Let's try to make this as clear as possible. 

Do we actually need to keep these numbers here? I was thinking about this again, could we instead do something like

Math.min(cipherBlockSize, remaining)?

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,212 @@
+      }
+
+      int remainingInBlock = plaintextBlock.length - currentOffsetInPlainBlock;
+      boolean finishTheBlock = remaining >= remainingInBlock;
```
Not sure I understand this? remainingInBlock is unencrypted bytes and remaining is encrypted right?

ggershinsky replied:
remaining is also unencrypted. If it's longer than what's left in current block, we'll read the block to the end

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,212 @@
+      if (endOfStream) {
+        break;
+      }
+      if (finishTheBlock) {
```
Could we change skip the break and just have it be 
```java
if (finishTheBlock && !endOfStream)
```

ggershinsky replied:
we need the break to stop the loop, no need to continue reading if the stream has ended.
finishTheBlock simply means we've finished one block, and start reading the next.

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,212 @@
+      remaining -= toCopy;
+      resultBufferOffset += toCopy;
+      currentOffsetInPlainBlock += toCopy;
+      boolean endOfStream = lastBlock && finishTheBlock;
```
This should also be remaining == 0 right?

ggershinsky replied:
remaining can be > 0 ; it just means the stream had less bytes than `len`, and the method returns ` len - remaining`

ggershinsky commented on core/src/test/java/org/apache/iceberg/encryption/TestGcmStreams.java:
```diff
@@ -0,0 +1,104 @@
+  public TemporaryFolder temp = new TemporaryFolder();
+
+  @Test
+  public void testRandomWriteRead() throws IOException {
```
Sure, will add

ggershinsky commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,212 @@
+    cipherBlockSize = plainBlockSize + AesGcmOutputStream.GCM_NONCE_LENGTH + AesGcmOutputStream.GCM_TAG_LENGTH;
+
+    try {
+      gcmCipher = Cipher.getInstance("AES/GCM/NoPadding");
```
Yep. We actually already have a Ciphers class; let's re-use it (with some additions) for all GCM-related things

ggershinsky commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,212 @@
+
+  @Override
+  public int available() throws IOException {
+    return Math.toIntExact(plainStreamSize - plainStreamPosition);
```
oh, no, streams typically use long for these, eg
```java
@Override
  public void seek(long newPos)
```

ggershinsky replied:
interesting though, the `available` doesn't have to be accurate, according to the InputStream contract. I presume this means that we don't need to use Math.toIntExact since it throws an exception if the result can't be mapped to an int. I'll change the logic to return max int in that case.

RussellSpitzer commented:
Let's try to finalize this before we get 0.14/ 1.0

ggershinsky commented:
SGTM. I've added a a commit with an optimization; no other changes are planned, so the PR is good to go from my POV.

RussellSpitzer reviewed (commented):
@jackye1995 + @rdblue Could you please take a look as well? Since this will define whole file encryption I was hoping we could solidify this before we do the major version bump.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputFile.java:
```diff
@@ -0,0 +1,71 @@
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.io.SeekableInputStream;
+
+public class AesGcmInputFile implements InputFile {
```
Why not implement `EncryptedInputFile`?

ggershinsky replied:
`EncryptedInputFile` is an input to encryption manager's `decrypt` method; with the output being an `InputFile`, ready for direct consumption by readers. `AesGcmInputFile` is an output of the decrypt operation (its input_stream.read draws decrypted data); so it is given directly to manifest Avro readers, that use it in the usual manner (decryption is transparent to them).

rdblue replied:
You're right. This would be the result of the `decrypt` method.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/Ciphers.java:
```diff
@@ -19,19 +19,29 @@
-  private static final int GCM_TAG_LENGTH = 16;
+  public static final int NONCE_LENGTH = 12;
+  public static final int GCM_TAG_LENGTH = 16;
+  public static final String GCM_STREAM_MAGIC_STRING = "GCM1";
```
Where does this come from? Is it introduced by Iceberg or from another standard?

ggershinsky replied:
The GCM Streams are introduced in this PR - being proposed as a part of the Iceberg spec / standard.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/Ciphers.java:
```diff
@@ -53,10 +63,14 @@ public AesGcmEncryptor(byte[] keyBytes) {
+      return encrypt(plaintext, 0, plaintext.length, aad);
+    }
+
+    public byte[] encrypt(byte[] plaintext, int plaintextOffset, int plaintextLength, byte[] aad) {
```
Where is this encryption format documented so I can take a look at compare it to the spec?

ggershinsky replied:
Yep, this (and other GCM Stream fields) are not documented yet. In a way, the GCM Streams are a simple file format, designed to add encryption and integrity verification to existing Input/Output Streams, in a transparent (and efficient) manner. I agree we need an explicit spec for this. Should I add a .md file (to the `iceberg/format` folder) in this PR so we can discuss it here?

ggershinsky replied:
@rdblue I've created a new PR, with the spec doc: https://github.com/apache/iceberg/pull/5432

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/Ciphers.java:
```diff
@@ -96,33 +110,45 @@ public AesGcmDecryptor(byte[] keyBytes) {
+      return decrypt(ciphertext, 0, ciphertext.length, aad);
+    }
+
+    public byte[] decrypt(byte[] ciphertext, int ciphertextOffset, int ciphertextLength, byte[] aad)  {
```
Should this be `decryptBlock` instead?

ggershinsky replied:
sure, we can rename it to `decryptBlock`

ggershinsky replied:
Unfortunately, this change triggers

> Execution failed for task ':iceberg-core:revapi'.
> > There were Java public API/ABI breaks reported by revapi:
>   java.method.removed: Method was removed.

How this should be handled?

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/Ciphers.java:
```diff
@@ -96,33 +110,45 @@ public AesGcmDecryptor(byte[] keyBytes) {
 
-      return plainText;
+  static byte[] streamBlockAAD(byte[] fileAadPrefix, int currentBlockIndex) {
+    byte[] blockAAD = Ints.toByteArray(currentBlockIndex);
```
I don't think this is specific enough. What is happening here? What is the endianness?

ggershinsky replied:
per the javadoc, https://github.com/google/guava/blob/master/guava/src/com/google/common/primitives/Ints.java#L300 ,
this is an explicitly big endian operation. But we can also spell it out in a comment above this line in our code - or create our own ~ `intToBytesBigEndian` function. -?

rdblue replied:
Why use big endian when other places use little endian?

ggershinsky replied:
Agreed, we'll use the same endianness.

ggershinsky commented:
Ack, I'm on it

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputFile.java:
```diff
@@ -0,0 +1,71 @@
+  public long getLength() {
+    if (plaintextLength == -1) {
+      try {
+        this.newStream().close();
```
We should be able to know the length without opening the stream, right?

ggershinsky replied:
Per our recent discussion, we'll set the default block size to 1MB, so the plaintext length can be calculated without opening the stream.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,218 @@
+
+    this.emptyCipherStream = (0 == netSourceFileSize);
+    this.sourceStream = sourceStream;
+    byte[] prefixBytes = new byte[Ciphers.GCM_STREAM_PREFIX_LENGTH];
```
Can we call this a header? We use `prefix` in several places and I think it would be more clear to call it a header since it is only at the start of the file.

ggershinsky replied:
SGTM

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,218 @@
+    int resultBufferOffset = off;
+    int remaining = len;
+
+    sourceStream.seek(Ciphers.GCM_STREAM_PREFIX_LENGTH + currentBlockIndex * cipherBlockSize);
```
It would be good to have some of these calculations as util methods. This could be `blockOffset(blockIndex)`.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,218 @@
+
+    boolean lastBlock = (currentBlockIndex + 1 == numberOfBlocks);
+    int resultBufferOffset = off;
+    int remaining = len;
```
Using `remaining` is a bit confusing because its meaning isn't clear. It could be remaining in the encrypted stream or remaining bytes to read, or something else. I'd recommend being more specific and using `bytesToRead`.

ggershinsky replied:
SGTM

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,218 @@
+                    byte[] aesKey, byte[] fileAadPrefix) throws IOException {
+    this.netSourceFileSize = sourceLength - Ciphers.GCM_STREAM_PREFIX_LENGTH;
+    Preconditions.checkArgument(netSourceFileSize >= 0,
+        "Source length " + sourceLength + " is shorter than GCM prefix. File is not encrypted");
```
If the unencrypted file length is 0, we should return an in-memory input stream with 0 bytes instead of a decrypting one.

ggershinsky replied:
This code can be reached only if the parent metadata has marked this file as encrypted (and provided us with the decryption key). If the file is not encrypted, this is an error situation.

rdblue replied:
If we know from the file size that the content is 0 bytes, we don't need to use an encrypted stream. I think we should fail if the header isn't there and something tries to read the file, but if there are no encrypted blocks then we can just return a simpler stream.

That will get rid of the `emptyCipherStream` cases and simplify this implementation.

ggershinsky replied:
Sounds reasonable, I'll apply this to the code.

ggershinsky replied:
I've tried this, and ran into the following issue. If the execution has reached this place, it means we expect an encrypted file (for example - reading manifests with non-null key_metadata). Maybe we should be stricter here, and simply throw an exception for empty cipher streams; these classes are designed for encryption of non-empty content; also, this will help to detect wrong situations earlier. What do you think?

ggershinsky commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,218 @@
+      numberOfBlocks = fullBlocksOnly ? numberOfFullBlocks : numberOfFullBlocks + 1;
+      lastCipherBlockSize = fullBlocksOnly ? cipherBlockSize : cipherBytesInLastBlock; // never 0
+      int plainBytesInLastBlock = fullBlocksOnly ? 0 :
+          (cipherBytesInLastBlock - Ciphers.NONCE_LENGTH - Ciphers.GCM_TAG_LENGTH);
```
This is the last block, so its length will be variable and shorter than the block size constant.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,218 @@
+    this.sourceStream = sourceStream;
+    byte[] prefixBytes = new byte[Ciphers.GCM_STREAM_PREFIX_LENGTH];
+    int fetched = sourceStream.read(prefixBytes);
+    Preconditions.checkState(fetched == Ciphers.GCM_STREAM_PREFIX_LENGTH,
```
This conclusion isn't correct. While unlikely to happen, the underlying stream is allowed to return fewer than the requested bytes. That isn't an error case and it doesn't indicate that the stream hit EOF.

If you want to ensure that you get all the bytes requested, then use `IOUtil.readFully`.

ggershinsky replied:
SGTM

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,218 @@
+    }
+  }
+
+  public long plaintextStreamSize() {
```
I think that this can be removed when the block size is set to a constant.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,218 @@
+    }
+
+    if (available() <= 0) {
+      return -1;
```
I think most Iceberg streams throw `EOFException` instead of returning -1.

ggershinsky replied:
I see, lets do the same here.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,218 @@
+  @Override
+  public int read(byte[] b, int off, int len) throws IOException {
+    if (len < 0) {
+      throw new IOException("Negative read length " + len);
```
I think this should be an `IllegalArgumentException`. Can you use a precondition?

ggershinsky replied:
SGTM

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,218 @@
+      return -1;
+    }
+
+    boolean lastBlock = (currentBlockIndex + 1 == numberOfBlocks);
```
Nit: name should be `isLastBlock`

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,218 @@
+      int toLoad = lastBlock ? lastCipherBlockSize : cipherBlockSize;
+      int loaded = sourceStream.read(ciphertextBlockBuffer, 0, toLoad);
+      if (loaded != toLoad) {
+        throw new IOException("Should read " + toLoad + " bytes, but got only " + loaded + " bytes");
```
This isn't an exceptional case. If you need to load the entire buffer, use `IOUtils.readFully`.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,218 @@
+    sourceStream.seek(Ciphers.GCM_STREAM_PREFIX_LENGTH + currentBlockIndex * cipherBlockSize);
+
+    while (remaining > 0) {
+      int toLoad = lastBlock ? lastCipherBlockSize : cipherBlockSize;
```
I find the naming a bit confusing because it is inconsistent. I think it is good to use `plain` and `cipher` to distinguish but we should do that consistently. In this case, I would call this `currentCipherBlockSize`.

I think it would also be good to break this up so you're not combining encrypted block reads and decrypts with the logic to copy data from a plaintext buffer. Can you add a `nextCipherBlock` or `nextPlainBlock` method that handles this?

ggershinsky replied:
SGTM

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,218 @@
+      }
+
+      byte[] aad = Ciphers.streamBlockAAD(fileAadPrefix, currentBlockIndex);
+      byte[] plaintextBlock = gcmDecryptor.decrypt(ciphertextBlockBuffer, 0, toLoad, aad);
```
It looks to me like this is going to read and decrypt the current block every time `read` is called. Am I wrong? Why not keep the current decrypted content in a buffer that persists across calls and consume bytes as needed from that buffer?

ggershinsky replied:
Yep, the current code is focused on manifest files, so I thought of adding this optimization to a later version for heavier  Avro data files. But I can add a single-block buffer (for the "current position block") now, shouldn't be too much extra complexity. Later, we can consider multi-block buffers and their management.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,274 @@
+      throw new IOException("Negative new position " + newPos);
+    } else if (newPos > plainStreamSize) {
+      throw new IOException(
+          "New position " + newPos + " exceeds the max stream size " + plainStreamSize);
```
Shouldn't this be EOFException?

ggershinsky replied:
SGTM.

rdblue commented on .palantir/revapi.yml:
```diff
@@ -743,6 +743,12 @@ acceptedBreaks:
+    - code: "java.method.visibilityReduced"
+      old: "method void org.apache.iceberg.encryption.Ciphers::<init>()"
+      new: "method void org.apache.iceberg.encryption.Ciphers::<init>()"
+      justification: "Static utility class - should not have public constructor"
```
Is this needed? Can we just leave it as it was or is it now failing?

ggershinsky replied:
Yep, the CI is failing unfortunately, since util classes are now required to have private constructors.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmOutputStream.java:
```diff
@@ -0,0 +1,116 @@
+import org.apache.iceberg.io.PositionOutputStream;
+
+public class AesGcmOutputStream extends PositionOutputStream {
+  public static final int plainBlockSize = 1024 * 1024;
```
`static final` variables should use names that are ALL_CAPS.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputFile.java:
```diff
@@ -0,0 +1,77 @@
+
+  @Override
+  public SeekableInputStream newStream() {
+    getLength();
```
Why is this called? It isn't used anywhere.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/Ciphers.java:
```diff
@@ -27,10 +30,19 @@
+  static final byte[] GCM_STREAM_MAGIC_ARRAY =
+      GCM_STREAM_MAGIC_STRING.getBytes(StandardCharsets.UTF_8);
+  static final int GCM_STREAM_HEADER_LENGTH =
+      GCM_STREAM_MAGIC_ARRAY.length + 4; // magic_len + block_size_len
```
I thought we had decided on a static block size?

ggershinsky replied:
correct. But its length is always 4.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/Ciphers.java:
```diff
@@ -67,7 +83,7 @@ public byte[] encrypt(byte[] plainText, byte[] aad) {
           cipher.updateAAD(aad);
         }
-        cipher.doFinal(plainText, 0, plainText.length, cipherText, NONCE_LENGTH);
+        cipher.doFinal(plaintext, plaintextOffset, plaintextLength, cipherText, NONCE_LENGTH);
```
Does the call to `getInstance` in the constructor guarantee that the `Cipher` instance is not shared? I'm guessing that it must if it has state (mode, key, etc.) but I don't see it guaranteed anywhere.

rdblue replied:
Should this check that the number of bytes returned by `doFinal` is correct?

rdblue replied:
Can you confirm that `doFinal` also writes the GCM tag to the output buffer?

ggershinsky replied:
> Does the call to getInstance in the constructor guarantee that the Cipher instance is not shared? 

The `AesGcmEncryptor` and `AesGcmDecryptor` are designed for a single thread use.

> Should this check that the number of bytes returned by doFinal is correct?

Sure, we can check this.

> Can you confirm that doFinal also writes the GCM tag to the output buffer?

From the `Cipher.doFinal` javadoc: 
 _If an AEAD mode such as GCM/CCM is being used, the authentication tag is appended in the case of encryption, or verified in the case of decryption._

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/Ciphers.java:
```diff
@@ -101,37 +117,52 @@ public AesGcmDecryptor(byte[] keyBytes) {
+        byte[] ciphertext, int ciphertextOffset, int ciphertextLength, byte[] aad) {
       Preconditions.checkState(
-          plainTextLength >= 1,
+          ciphertextLength - GCM_TAG_LENGTH - NONCE_LENGTH >= 1,
```
Should there be a similar check in the encrypt method to validate that the plaintext has at least one byte?

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/Ciphers.java:
```diff
@@ -101,37 +117,52 @@ public AesGcmDecryptor(byte[] keyBytes) {
       // Get the nonce from ciphertext
       byte[] nonce = new byte[NONCE_LENGTH];
-      System.arraycopy(ciphertext, 0, nonce, 0, NONCE_LENGTH);
+      System.arraycopy(ciphertext, ciphertextOffset, nonce, 0, NONCE_LENGTH);
```
The allocation here is unnecessary. `GCMParameterSpec` accepts an offset and length for the nonce array.

ggershinsky replied:
Cool!

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/Ciphers.java:
```diff
@@ -101,37 +117,52 @@ public AesGcmDecryptor(byte[] keyBytes) {
 
-      byte[] plainText = new byte[plainTextLength];
-      int inputLength = ciphertext.length - NONCE_LENGTH;
+      int inputLength = ciphertextLength - NONCE_LENGTH;
```
This looks okay, but I'd want to double check how the tag bytes are handled. Here, the length includes the tag bytes so this isn't the same as the plaintext length that is used for encryption.

ggershinsky replied:
Correct. 
I removed this var, its name is not intuitive enough.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/Ciphers.java:
```diff
@@ -101,37 +117,52 @@ public AesGcmDecryptor(byte[] keyBytes) {
+      return decrypt(ciphertext, 0, ciphertext.length, aad);
+    }
+
+    public byte[] decrypt(
```
I think that `TestCiphers` needs to have tests validating failure modes:
* If the supplied AAD is incorrect, decryption fails
* If the ciphertext is modified, decryption fails

Right now, only the happy path is tested, which doesn't validate the intended authentication behavior using the tag.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,274 @@
+    return (long) blockIndex * cipherBlockSize + Ciphers.GCM_STREAM_HEADER_LENGTH;
+  }
+
+  private static long netSourceFileLength(long sourceFileLength) {
```
I don't think that "net" is clear enough, and I also think that this method is doing too much. A calculation should not include a Precondition check like this because it is surprising. You must know the implementation details of the method to call it properly.

I think that the length should be validated in `newStream` when attempting to read the file, and _not_ when calculating the file length for `InputStream.getLength`.

I think checking at that point avoids the need to have this method entirely, which avoids the naming problem. The call can simply be `encryptedFileLength - Ciphers.GCM_STREAM_HEADER_LENGTH`.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,274 @@
+  }
+
+  static long calculatePlaintextLength(long sourceLength, int plainBlockSize) {
+    long netSourceFileLength = netSourceFileLength(sourceLength);
```
As I noted below, I don't think "net" is very clear. How about `long streamLength = sourceLength - Ciphers.GCM_STREAM_HEADER_LENGTH`?

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,274 @@
+    long netSourceFileLength = netSourceFileLength(sourceLength);
+
+    if (netSourceFileLength == 0) {
+      return 0;
```
I think this means that you can have a 0-length content file. That might make sense, but if we are going to allow a 0-length file then it should actually use a 0-length block as well.

The problem with a 0-length stream is that an attacker can replace a file with a 0-length stream to effectively delete data while bypassing authentication checks. A 0-length block would prevent that attack because it would check the AAD for the block.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,274 @@
+      return 0;
+    }
+
+    int cipherBlockSize = plainBlockSize + Ciphers.NONCE_LENGTH + Ciphers.GCM_TAG_LENGTH;
```
I think that this should be a constant and that `plainBlockSize` should not be passed in. I realize that the purpose is to make this generic if we ever want to support another block size, but all it does right now is make the code more confusing. I'm okay keeping the block size in the header, but we should not pass it around in the code. We can always update the code later.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,274 @@
+    int cipherBlockSize = plainBlockSize + Ciphers.NONCE_LENGTH + Ciphers.GCM_TAG_LENGTH;
+    int numberOfFullBlocks = Math.toIntExact(netSourceFileLength / cipherBlockSize);
+    int cipherBytesInLastBlock =
+        Math.toIntExact(netSourceFileLength - numberOfFullBlocks * cipherBlockSize);
```
Since the stream length is a long, I don't think it makes sense to make `numberOfFullBlocks` in int. The multiplication can clearly produce a long anyway so this should remove the `toIntExact` above. Then everything is a long until `cipherBytesInLastBlock`.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,274 @@
+            ? 0
+            : (cipherBytesInLastBlock - Ciphers.NONCE_LENGTH - Ciphers.GCM_TAG_LENGTH);
+
+    return (long) numberOfFullBlocks * plainBlockSize + plainBytesInLastBlock;
```
Using a `long` for `numberOfFullBlocks` avoids the need for a cast here. Also, since this is producing a `long`, I don't think it makes sense to do the extra logic to keep either `cipherBytesInLastBlock` or `plainBytesInLastBlock` as integers. Just use longs and not worry about the casts.

rdblue replied:
Here's the full method with my changes to types:

```java
  static long calculatePlaintextLength(long sourceLength) {
    long streamLength = sourceLength - Ciphers.GCM_STREAM_HEADER_LENGTH;

    if (streamLength == 0) {
      return 0;
    }

    long numberOfFullBlocks = streamLength / AesGcmOutputStream.CIPHER_BLOCK_SIZE;
    long fullBlockSize = numberOfFullBlocks * AesGcmOutputStream.CIPHER_BLOCK_SIZE;
    long cipherBytesInLastBlock = streamLength - fullBlockSize;
    boolean fullBlocksOnly = (0 == cipherBytesInLastBlock);
    long plainBytesInLastBlock =
        fullBlocksOnly
            ? 0
            : (cipherBytesInLastBlock - Ciphers.NONCE_LENGTH - Ciphers.GCM_TAG_LENGTH);

    return (numberOfFullBlocks * AesGcmOutputStream.PLAIN_BLOCK_SIZE) + plainBytesInLastBlock;
  }
```

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,274 @@
+
+  @Override
+  public int read() throws IOException {
+    throw new UnsupportedOperationException();
```
This is required.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,274 @@
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+    return n;
```
Style: missing newline after the previous control flow block.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,274 @@
+    this.sourceStream = sourceStream;
+    byte[] headerBytes = new byte[Ciphers.GCM_STREAM_HEADER_LENGTH];
+    IOUtil.readFully(sourceStream, headerBytes, 0, headerBytes.length);
+    byte[] magic = new byte[Ciphers.GCM_STREAM_MAGIC_ARRAY.length];
```
This allocation isn't needed. You can wrap the first 4 bytes of the header array in a ByteBuffer and use that for the comparison. That is also better to that you don't need to expose a mutable array in Ciphers:

```java
  static final ByteBuffer GCM_STREAM_MAGIC = ByteBuffer.wrap(GCM_STREAM_MAGIC_ARRAY).asReadOnlyBuffer();
```

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,274 @@
+    boolean emptyCipherStream = (0 == netSourceLength);
+    this.sourceStream = sourceStream;
+    byte[] headerBytes = new byte[Ciphers.GCM_STREAM_HEADER_LENGTH];
+    IOUtil.readFully(sourceStream, headerBytes, 0, headerBytes.length);
```
This should avoid reading in the constructor. Otherwise, this needs to throw IOException and it is harder for the caller to ensure that resources are closed if anything goes wrong.

For example, if this throws `IOException` then `newStream` isn't going to return anything. This stream is never constructed, but this is the owner of `sourceFile.newStream()` that is passed in as `sourceStream`. As a result, `sourceStream` is leaked.

Instead, this should lazily initialize and use something like the `currentBlockIndex` to keep track of whether it needs to init.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,274 @@
+          ByteBuffer.wrap(headerBytes, Ciphers.GCM_STREAM_MAGIC_ARRAY.length, 4)
+              .order(ByteOrder.LITTLE_ENDIAN)
+              .getInt();
+      Preconditions.checkState(plainBlockSize > 0, "Wrong plainBlockSize " + plainBlockSize);
```
This duplicates the next check.

rdblue commented:
@ggershinsky, I had a lot of comments, so I went ahead and implemented most of them on top of this. I think it simplified how the input stream works quite a bit and I fixed a lot of the comments that I made. Please see https://github.com/ggershinsky/iceberg/pull/6.

I think we'll also need to apply similar changes to the output stream, but I haven't had a chance to take a close look yet.

ggershinsky commented:
> I had a lot of comments, so I went ahead and implemented most of them on top of this. I think it simplified how the input stream works quite a bit and I fixed a lot of the comments that I made. Please see https://github.com/ggershinsky/iceberg/pull/6.

Thanks! Merged the PR.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java:
```diff
@@ -0,0 +1,264 @@
+  }
+
+  private int availableInCurrentBlock() {
+    if (currentPlainBlockIndex < 0) {
```
When I was refactoring, there was a bug that I fixed when the block for `plainStreamPosition` didn't match `currentPlainBlockIndex`. I ended up fixing the problem by setting `currentPlainBlockIndex = -1` in a place that I had missed. Now that I'm thinking about it more, I think the right solution is actually to update the check here instead. That way the bytes available will only be non-zero if the current block matches the position.

Here's a diff that does what I'm talking about and still passes tests:

```diff
diff --git a/core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java b/core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java
index a63134f31d..88e9b36c25 100644
--- a/core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java
+++ b/core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java
@@ -97,7 +97,7 @@ public class AesGcmInputStream extends SeekableInputStream {
   }
 
   private int availableInCurrentBlock() {
-    if (currentPlainBlockIndex < 0) {
+    if (blockIndex(plainStreamPosition) != currentPlainBlockIndex) {
       return 0;
     }
 
@@ -130,10 +130,6 @@ public class AesGcmInputStream extends SeekableInputStream {
         remainingBytesToRead -= bytesToCopy;
         resultBufferOffset += bytesToCopy;
         this.plainStreamPosition += bytesToCopy;
-        if (blockIndex(plainStreamPosition) != currentPlainBlockIndex) {
-          // invalidate the current block
-          this.currentPlainBlockIndex = -1;
-        }
 
       } else if (available() > 0) {
         decryptBlock(blockIndex(plainStreamPosition));
@@ -157,10 +153,6 @@ public class AesGcmInputStream extends SeekableInputStream {
     }
 
     this.plainStreamPosition = newPos;
-    if (blockIndex(plainStreamPosition) != currentPlainBlockIndex) {
-      // invalidate the current block
-      this.currentPlainBlockIndex = -1;
-    }
   }
 
   @Override
@@ -177,10 +169,6 @@ public class AesGcmInputStream extends SeekableInputStream {
     }
 
     this.plainStreamPosition += n;
-    if (blockIndex(plainStreamPosition) != currentPlainBlockIndex) {
-      // invalidate the current block
-      this.currentPlainBlockIndex = -1;
-    }
 
     return n;
   }
```

ggershinsky replied:
SGTM. Applied this change.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmOutputFile.java:
```diff
@@ -0,0 +1,66 @@
+  public PositionOutputStream create() {
+    try {
+      return new AesGcmOutputStream(targetFile.create(), dataKey, fileAADPrefix);
+    } catch (IOException e) {
```
Similar to my [note above](https://github.com/apache/iceberg/pull/3231/files#r1278605737), I don't think that constructors should throw IOException.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmOutputFile.java:
```diff
@@ -0,0 +1,66 @@
+
+  @Override
+  public InputFile toInputFile() {
+    throw new UnsupportedOperationException();
```
This should be implemented.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmOutputStream.java:
```diff
@@ -0,0 +1,115 @@
+    this.gcmEncryptor = new Ciphers.AesGcmEncryptor(aesKey);
+    this.plainBlockBuffer = new byte[Ciphers.PLAIN_BLOCK_SIZE];
+    this.positionInBuffer = 0;
+    this.streamPosition = 0;
```
Shouldn't this always be `targetStream.pos() + positionInBuffer`?

ggershinsky replied:
this value is returned by 

```java
@Override
  public long getPos() throws IOException {
    return streamPosition;
  }
```

  so is not related directly to targetStream.pos() and positionInBuffer.
I'll rename `positionInBuffer` to `positionInPlainBlock`, and `plainBlockBuffer` to `plainBlock`, to make the naming more consistent.

rdblue replied:
I don't see why what I said wouldn't always be true. If so, we could eliminate this extra state.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmOutputStream.java:
```diff
@@ -0,0 +1,115 @@
+            .put(Ciphers.GCM_STREAM_MAGIC_ARRAY)
+            .putInt(Ciphers.PLAIN_BLOCK_SIZE)
+            .array();
+    targetStream.write(headerBytes);
```
This should be done outside of the constructor before writing the first block.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmOutputStream.java:
```diff
@@ -0,0 +1,115 @@
+
+  @Override
+  public void write(int b) throws IOException {
+    throw new UnsupportedOperationException();
```
I think this needs to be implemented as well.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmOutputStream.java:
```diff
@@ -0,0 +1,115 @@
+      throw new IOException(
+          "Insufficient bytes in buffer: " + b.length + " - " + off + " < " + len);
+    }
+    int remaining = len;
```
Nit: missing newline above this.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmOutputStream.java:
```diff
@@ -0,0 +1,115 @@
+    int offset = off;
+
+    while (remaining > 0) {
+      int freeBlockBytes = Ciphers.PLAIN_BLOCK_SIZE - positionInBuffer;
```
While it should be the same, I think that this should use `plainBlockBuffer.length` instead of `Ciphers.PLAIN_BLOCK_SIZE` because that's the length that will avoid an `ArrayIndexOutOfBoundsException`. Otherwise, we're relying on the assumption that `plainBlockBuffer.length == Ciphers.PLAIN_BLOCK_SIZE` that may not hold later.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmOutputStream.java:
```diff
@@ -0,0 +1,115 @@
+
+    while (remaining > 0) {
+      int freeBlockBytes = Ciphers.PLAIN_BLOCK_SIZE - positionInBuffer;
+      int toWrite = freeBlockBytes <= remaining ? freeBlockBytes : remaining;
```
I think it's better to use `Math.min(freeBlockBytes, remaining)`.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmOutputStream.java:
```diff
@@ -0,0 +1,115 @@
+        encryptAndWriteBlock();
+        positionInBuffer = 0;
+      }
+      offset += toWrite;
```
I think it would be better to do all of the offset accounting in the same place, rather than breaking it up with the check for a full block.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmOutputStream.java:
```diff
@@ -0,0 +1,115 @@
+
+      System.arraycopy(b, offset, plainBlockBuffer, positionInBuffer, toWrite);
+      positionInBuffer += toWrite;
+      if (positionInBuffer == Ciphers.PLAIN_BLOCK_SIZE) {
```
You'll probably want to move this to a method that can be called from `write(int)` like `flushBlock()`.

ggershinsky replied:
I've implemented `write(int)` by calling `write(byte[])`, which in turn calls this method (`write(byte[], offset, len)`). So all of accounting and encryption is in one place.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmOutputStream.java:
```diff
@@ -0,0 +1,115 @@
+    targetStream.close();
+  }
+
+  private void encryptAndWriteBlock() throws IOException {
```
Do you think it would be good to have a boolean `fullBlock` to keep track of whether the last block was a full block? Then we could check it here and refuse to write another block if the last block was not full.

ggershinsky replied:
this function is called only for full blocks, after an explicit check:

```java
if (positionInPlainBlock == Ciphers.PLAIN_BLOCK_SIZE) {
        encryptAndWriteBlock();
```

rdblue replied:
This is really just to sanity check the state and avoid errors. The code is correct today because your assumption holds, but as this class changes, we want to make as many assumptions as possible explicit rather than relying on people knowing what to do.

ggershinsky replied:
In my previous comment, I forgot about another call of this function - from the stream `close()` method. There, the last block is not assumed to be full, since this writes the remains of the plainBlock buffer.

ggershinsky replied:
I've added a check making sure the block is not empty

rdblue replied:
Actually, I meant that it would be good to have a check that only one partial block (last block) is written.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/Ciphers.java:
```diff
@@ -55,10 +71,15 @@ public AesGcmEncryptor(byte[] keyBytes) {
       byte[] nonce = new byte[NONCE_LENGTH];
       randomGenerator.nextBytes(nonce);
-      int cipherTextLength = NONCE_LENGTH + plainText.length + GCM_TAG_LENGTH;
+      int cipherTextLength = NONCE_LENGTH + plaintextLength + GCM_TAG_LENGTH;
```
Is this buffer allocation necessary? Why can't we pass this in and reuse it for every block? It is passed into the `doFinal` call so I think we could have this return the number of bytes that were encrypted and reuse a buffer from the caller.

ggershinsky replied:
Done for simplicity, since manifest encryption won't benefit much from this optimization. 
But we can add now, for Avro data encryption in the future.
Btw, the same optimization can be performed for decryption.

ggershinsky replied:
Added to both encryption and decryption in the new commit.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmOutputStream.java:
```diff
@@ -0,0 +1,115 @@
+    byte[] aad = Ciphers.streamBlockAAD(fileAadPrefix, currentBlockIndex);
+    byte[] cipherBlockBuffer = gcmEncryptor.encrypt(plainBlockBuffer, 0, positionInBuffer, aad);
+    currentBlockIndex++;
+    targetStream.write(cipherBlockBuffer);
```
I think that this should reset `positionInBuffer` instead of doing it in the caller.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmOutputStream.java:
```diff
@@ -0,0 +1,138 @@
+            .order(ByteOrder.LITTLE_ENDIAN)
+            .put(Ciphers.GCM_STREAM_MAGIC_ARRAY)
+            .putInt(Ciphers.PLAIN_BLOCK_SIZE)
+            .array();
```
This byte array is made entirely from constants. Can you make it a constant byte array?

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/AesGcmOutputStream.java:
```diff
@@ -0,0 +1,138 @@
+
+  private void encryptAndWriteBlock() throws IOException {
+    if (currentBlockIndex == Integer.MAX_VALUE) {
+      throw new IOException("Too many blocks - exceed Integer.MAX_VALUE");
```
I think this should be `Cannot write block: exceeded Integer.MAX_VALUE blocks`.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/Ciphers.java:
```diff
@@ -53,29 +70,61 @@ public AesGcmEncryptor(byte[] keyBytes) {
+        byte[] ciphertextBuffer,
+        int ciphertextOffset,
+        byte[] aad) {
+      Preconditions.checkArgument(plaintextLength > 0, "Wrong plaintextLength " + plaintextLength);
```
We typically use `Invalid plain text length: %s`:
* "Invalid" is used more often and signals that the value cannot be used, while "wrong" could indicate that the value was not equal to some expected value or failed a different type of check.
* Variable names like `plaintextLength` aren't helpful to people reading error messages because they aren't familiar with this code. Using plain English to describe the problem is better.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/Ciphers.java:
```diff
@@ -53,29 +70,61 @@ public AesGcmEncryptor(byte[] keyBytes) {
           cipher.updateAAD(aad);
         }
-        cipher.doFinal(plainText, 0, plainText.length, cipherText, NONCE_LENGTH);
+        enciphered =
```
Style: needs a newline before this.

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/Ciphers.java:
```diff
@@ -53,29 +70,61 @@ public AesGcmEncryptor(byte[] keyBytes) {
+                  + enciphered
+                  + ". Must be "
+                  + plaintextLength
+                  + GCM_TAG_LENGTH);
```
I think this can be more clear because it doesn't tell the reader that the encryption didn't work right.

Instead, how about `Failed to encrypt block: expected %s encrypted bytes but produced %s bytes`?

rdblue commented on core/src/main/java/org/apache/iceberg/encryption/Ciphers.java:
```diff
@@ -53,29 +70,61 @@ public AesGcmEncryptor(byte[] keyBytes) {
+                plaintextOffset,
+                plaintextLength,
+                ciphertextBuffer,
+                ciphertextOffset + NONCE_LENGTH);
```
I don't see where the `nonce` is copied into the output `cyphertextBuffer`. Is that done inside of `doFinal`? If so, can you add a comment that states what exactly `doFinal` puts into the output buffer?

ggershinsky replied:
This is done outside `doFinal`, I'll add a comment.

rdblue commented on core/src/test/java/org/apache/iceberg/encryption/TestCiphers.java:
```diff
@@ -51,6 +53,21 @@ private void testEncryptDecrypt(byte[] aad) {
       byte[] decryptedText = decryptor.decrypt(ciphertext, aad);
       assertThat(decryptedText).as("Key length " + keyLength).isEqualTo(plaintext);
+
+      // Test bad aad
```
This needs to be a separate test case / method.

When you add additional cases to an existing test method, the tests are harder to work with because you don't get accurate descriptions of what failed. Plus, if there are issues with the case above, this test code won't be run so the person trying to figure out why tests are failing gets less information. Multiple failing cases are helpful for determining what needs to be fixed and for finding a good fix the first time that doesn't cause other failures.

rdblue commented:
@ggershinsky, I opened https://github.com/ggershinsky/iceberg/pull/8 with the remaining changes for this PR.

The main change is to write an empty block for empty files. Otherwise an attacker could replace a valid file with an empty file without detection.
